x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java

```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.core.UpdateForV9;
+import org.elasticsearch.ingest.EnterpriseGeoIpTask.EnterpriseGeoIpTaskParams;
+import org.elasticsearch.license.License;
+import org.elasticsearch.license.LicenseStateListener;
+import org.elasticsearch.license.LicensedFeature;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.persistent.PersistentTasksService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.RemoteTransportException;
+import org.elasticsearch.xpack.core.XPackField;
+
+import java.util.Objects;
+
+import static org.elasticsearch.ingest.EnterpriseGeoIpTask.ENTERPRISE_GEOIP_DOWNLOADER;
+
+public class EnterpriseGeoIpDownloaderLicenseListener implements LicenseStateListener, ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(EnterpriseGeoIpDownloaderLicenseListener.class);
+    // Note: This custom type is GeoIpMetadata.TYPE, but that class is not exposed to this plugin
+    static final String INGEST_GEOIP_CUSTOM_METADATA_TYPE = "ingest_geoip";
+
+    private final PersistentTasksService persistentTasksService;
+    private final ClusterService clusterService;
+    private final XPackLicenseState licenseState;
+    private static final LicensedFeature.Momentary ENTERPRISE_GEOIP_FEATURE = LicensedFeature.momentary(
+        null,
+        XPackField.ENTERPRISE_GEOIP_DOWNLOADER,
+        License.OperationMode.PLATINUM
+    );
+    private volatile boolean licenseIsValid = false;
+    private volatile boolean hasIngestGeoIpMetadata = false;
+
+    protected EnterpriseGeoIpDownloaderLicenseListener(
+        Client client,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        XPackLicenseState licenseState
+    ) {
+        this.persistentTasksService = new PersistentTasksService(clusterService, threadPool, client);
+        this.clusterService = clusterService;
+        this.licenseState = licenseState;
+    }
+
+    @UpdateForV9 // use MINUS_ONE once that means no timeout
+    private static final TimeValue MASTER_TIMEOUT = TimeValue.MAX_VALUE;
+    private volatile boolean licenseStateListenerRegistered;
+
+    public void init() {
+        listenForLicenseStateChanges();
+        clusterService.addListener(this);
+    }
+
+    void listenForLicenseStateChanges() {
+        assert licenseStateListenerRegistered == false;
+        licenseState.addListener(this);
+        licenseStateListenerRegistered = true;
```

Comment:
I think we might want to swap the order of these two, otherwise we could potentially hit the assert in `licenseStateChanged`. E.g., if the listener were added, then the were to pause before setting the value to `true`, and then `licenseStateChanged` got called from another thread prior to this value being updated.

What do you think?

Reply:
Now that it's only used to make sure we don't call init() twice (see my comment about the related assertion below), I don't think it matters. I'll move it anyway though.